flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java

```diff
@@ -0,0 +1,271 @@
+			throw new DatabaseNotExistException(catalogName, databaseName);
+		}
+	}
+
+	private boolean isDatabaseEmpty(String databaseName) {
+		checkArgument(!StringUtils.isNullOrWhitespaceOnly(databaseName));
+
+		return tables.keySet().stream().noneMatch(op -> op.getDatabaseName().equals(databaseName));
+		// TODO: also check function when function is added.
+	}
+
+	@Override
+	public void alterDatabase(String databaseName, CatalogDatabase newDatabase, boolean ignoreIfNotExists)
+		throws DatabaseNotExistException {
+		checkArgument(!StringUtils.isNullOrWhitespaceOnly(databaseName));
+		checkArgument(newDatabase != null);
+
+		if (databaseExists(databaseName)) {
+			databases.put(databaseName, newDatabase.copy());
+		} else if (!ignoreIfNotExists) {
+			throw new DatabaseNotExistException(catalogName, databaseName);
+		}
+	}
+
+	@Override
+	public List<String> listDatabases() {
+		return new ArrayList<>(databases.keySet());
+	}
+
+	@Override
+	public CatalogDatabase getDatabase(String databaseName) throws DatabaseNotExistException {
+		checkArgument(!StringUtils.isNullOrWhitespaceOnly(databaseName));
+
+		if (!databaseExists(databaseName)) {
+			throw new DatabaseNotExistException(catalogName, databaseName);
+		} else {
+			return databases.get(databaseName).copy();
+		}
+	}
+
+	@Override
+	public boolean databaseExists(String databaseName) {
+		checkArgument(!StringUtils.isNullOrWhitespaceOnly(databaseName));
+
+		return databases.containsKey(databaseName);
+	}
+
+	// ------ tables ------
+
+	@Override
+	public void createTable(ObjectPath tablePath, CommonTable table, boolean ignoreIfExists)
+		throws TableAlreadyExistException, DatabaseNotExistException {
+		checkArgument(tablePath != null);
+		checkArgument(table != null);
+
+		if (!databaseExists(tablePath.getDatabaseName())) {
+			throw new DatabaseNotExistException(catalogName, tablePath.getDatabaseName());
+		}
+
+		if (tableExists(tablePath)) {
+			if (!ignoreIfExists) {
+				throw new TableAlreadyExistException(catalogName, tablePath);
+			}
+		} else {
+			tables.put(tablePath, table.copy());
+		}
+	}
+
+	@Override
+	public void alterTable(ObjectPath tablePath, CommonTable newTable, boolean ignoreIfNotExists)
+		throws TableNotExistException {
+		checkArgument(tablePath != null);
+		checkArgument(newTable != null);
+
+		if (tableExists(tablePath)) {
+			tables.put(tablePath, newTable.copy());
+		} else if (!ignoreIfNotExists) {
+			throw new TableNotExistException(catalogName, tablePath);
+		}
+	}
+
+	// ------ tables and views ------
+
+	@Override
+	public void dropTable(ObjectPath tablePath, boolean ignoreIfNotExists) throws TableNotExistException {
+		checkArgument(tablePath != null);
+
+		if (tableExists(tablePath)) {
+			tables.remove(tablePath);
+		} else if (!ignoreIfNotExists) {
+			throw new TableNotExistException(catalogName, tablePath);
+		}
+	}
+
+	@Override
+	public void renameTable(ObjectPath tablePath, String newTableName, boolean ignoreIfNotExists)
```

Comment:
Can't we relax this requirement and allow moving between databases? I know some catalogs don't support that, but others might.

Reply:
Yes, we can. However, I'd think pure renaming is probably more common than moving, which is the reason why it's here. For moving table between databases, we may provide a different API purely for that purpose rather than overloading it here, if that requirement ever comes. Any further thoughts?

Reply:
Ok, make sense.